### PR TITLE
make repo as installable package

### PR DESCRIPTION
**Pull request by Borda**
I found it will be useful to have this project as an installable package so I suggest the following changes:
- rename actual `perceiver` as `model` (as it holds all model-related components)
- create new package `perceiverio` which pulls `model`, `data` and `cli` as sub-packages
- simplify the `cli` package as just module as it quite lite
- add setup.py to be installed

### CODE REVIEW

**Comment by Borda** [comment]
@krasserm could you pls have a look? :rabbit:

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+pytorch==1.10.*
```
Comment (krasserm) [inline requirements.txt:1]: Must be renamed to `torch`

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+torchvision==0.11.*
+torchtext==0.11.*
+fairscale==0.4
```
Comment (krasserm) [inline requirements.txt:4]: Should be `fairscale==0.4.*`

File: perceiverio/cli.py [hunk 3]
```diff
@@ -9,13 +9,11 @@
     LightningArgumentParser,
 )
 
-from cli.trainer import DDPStaticGraphPlugin
-
```
Comment (krasserm) [inline perceiverio/cli.py:12]: `from pytorch_lightning.plugins import DDPPlugin` is missing here

File: environment.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: perceiver-io
+name: model-io
```
Comment (krasserm) [inline environment.yml:1]: Keep environment name `perceiver-io`?

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    include_package_data=True,
+    classifiers=[
+        'Environment :: Console',
+        'Natural Language :: English',
+        # How mature is this project? Common values are
+        #   3 - Alpha, 4 - Beta, 5 - Production/Stable
+        'Development Status :: 4 - Beta',
+        # Indicate who your project is intended for
+        'Intended Audience :: Developers',
+        'Topic :: Software Development :: Build Tools',
```
Comment (krasserm) [inline setup.py:47]: What about using these topics instead?

```
'Topic :: Scientific/Engineering',
'Topic :: Scientific/Engineering :: Artificial Intelligence',
```

**Review by krasserm — COMMENTED** [review]
Thanks a lot for your great contribution @Borda, definitely a useful for this project! In addition to my previous comments I propose the following changes:

- Rename package `perceiverio` to `perceiver`
- Rename module `modules.py` (previously `model.py`) to `core.py`
- Imports in `notebooks/*.ipynb` need to be adjusted
- Reuse `requirements.txt` in `environment.yml`

I've already made these changes locally. If you don't mind I would add them to this branch for further review/discussion. Please let me know!

File: setup.py (hunk 5 above)
Comment (Borda) [reply setup.py:47]:
```suggestion
        'Topic :: Scientific/Engineering',
        'Topic :: Scientific/Engineering :: Artificial Intelligence
```

File: environment.yml (hunk 4 above)
Comment (Borda) [reply environment.yml:1]: I would say that model io is too general and doe not imply any connection to perceiver, but if you wish so :rabbit:

File: requirements.txt (hunk 2 above)
Comment (Borda) [reply requirements.txt:4]: good point!
```suggestion
fairscale==0.4.*
```

File: requirements.txt (hunk 1 above)
Comment (Borda) [reply requirements.txt:1]:
```suggestion
torch==1.10.*
```

File: README.md [hunk 6]
```diff
@@ -13,12 +13,19 @@ output adapters (see [Model API](#model-api)). The command line interface is imp
 
 ## Setup
 
+Configure with source code and Conda
+
 ```shell
 conda env create -f environment.yml
-conda activate perceiver-io
+conda activate model-io
```
Comment (Borda) [inline README.md:20]:
```suggestion
conda activate perceiver-io
```

File: environment.yml (hunk 4 above)
Comment (Borda) [reply environment.yml:1]:
```suggestion
name: perceiver-io
```

File: perceiverio/cli.py [hunk 7]
```diff
@@ -9,13 +9,11 @@
     LightningArgumentParser,
 )
```
Comment (Borda) [inline perceiverio/cli.py:11]:
```suggestion
from pytorch_lightning.plugins import DDPPlugin

```

File: perceiverio/cli.py (hunk 3 above)
Comment (Borda) [reply perceiverio/cli.py:13]: yep, thx

**Comment by Borda** [comment]
> * Rename package `perceiverio` to `perceiver`
> * Rename module `modules.py` (previously `model.py`) to `core.py`
> * Imports in `notebooks/*.ipynb` need to be adjusted

done :rabbit: 

> * Reuse `requirements.txt` in `environment.yml`

not sure if I understand what you mean here, if you suggest adding `-r requirements.txt` to the `environment.yml`, then I would rather drop the `environment.yml` completely :angel:

**Comment by krasserm** [comment]
> not sure if I understand what you mean here, if you suggest adding `-r requirements.txt` to the `environment.yml`, then I would rather drop the `environment.yml` completely :angel: 

Yes, this is what I mean. I still find it useful having a conda `environment.yml` for installing `cudatoolkit` (for users who do not have it installed on host OS or in another virtual environment).

File: .github/workflows/ci_install-pkg.yml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+          # this is just a hotfix because of Win cannot install it directly
+          pip install -r requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
+
+      - name: Install | Uninstall package - archive
+        run: |
+          # install as archive
+          pip install dist/*.tar.gz
+          cd ..
+          python -c "import perceiverio ; print(perceiverio.__version__)"
+          pip uninstall -y perceiver-io
```
Comment (krasserm) [inline .github/workflows/ci_install-pkg.yml:72]: `perceiver-io` and `perceiverio` escaped renaming. Should be `perceiver`.

File: .github/workflows/ci_install-pkg.yml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+      - name: Install | Uninstall package - wheel
+        run: |
+          # install as wheel
+          pip install dist/*.whl
+          cd ..
+          python -c "import perceiverio ; print(perceiverio.__version__)"
+          pip uninstall -y perceiver-io
```
Comment (krasserm) [inline .github/workflows/ci_install-pkg.yml:80]: Same here ...

File: MANIFEST.in [hunk 10]
```diff
@@ -0,0 +1,10 @@
+include *.txt
+exclude *.yml
+include LICENSE
+recursive-include docs *.md
+recursive-include docs *.png
+recursive-exclude notebooks *.ipynb
+recursive-include perceiverio *.yaml
```
Comment (krasserm) [inline MANIFEST.in:7]: Should be `perceiver`

**Review by krasserm — COMMENTED** [review]
Only two minor naming issues left. Otherwise, looks good to me!

File: .github/workflows/ci_install-pkg.yml (hunk 8 above)
Comment (Borda) [reply .github/workflows/ci_install-pkg.yml:72]: personally, I would keep it a bit different as package `scikit-learn` and import pkg is `sklearn`
note that they are two different things, the import is `import perceiver` but for install from pip or listed in packages with `pip list` it is `perceiver-io`

File: MANIFEST.in (hunk 10 above)
Comment (Borda) [reply MANIFEST.in:7]:
```suggestion
recursive-include perceiver *.yaml
```
good catch

File: .github/workflows/ci_install-pkg.yml (hunk 9 above)
Comment (Borda) [reply .github/workflows/ci_install-pkg.yml:80]: as above, but if you wich, can be unified :rabbit:

**Comment by Borda** [comment]
> > not sure if I understand what you mean here, if you suggest adding `-r requirements.txt` to the `environment.yml`, then I would rather drop the `environment.yml` completely angel
> 
> Yes, this is what I mean. I still find it useful having a conda `environment.yml` for installing `cudatoolkit` (for users who do not have it installed on host OS or in another virtual environment).

I see and it is good, the only thing I a not sure about is how Conda handle duplicates, cor example if you have `pytorch` in Conda and are referred again as `torch` in requirements.txt
If you would not mind I would leave this decision to the other open PR #4

File: .github/workflows/ci_install-pkg.yml (hunk 8 above)
Comment (krasserm) [reply .github/workflows/ci_install-pkg.yml:72]: I see, let's use `perceiver-io` then.

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    classifiers=[
+        'Environment :: Console',
+        'Natural Language :: English',
+        # How mature is this project? Common values are
+        #   3 - Alpha, 4 - Beta, 5 - Production/Stable
+        'Development Status :: 4 - Beta',
+        # Indicate who your project is intended for
+        'Intended Audience :: Developers',
+        'Topic :: Scientific/Engineering',
+        'Topic :: Scientific/Engineering :: Artificial Intelligence
```
Comment (krasserm) [inline setup.py:48]: There is a `',` missing at the end of this line ...

Comment (krasserm) [reply setup.py:48]: Sorry, missed that during my previous review ...

Comment (Borda) [reply setup.py:48]:
```suggestion
        'Topic :: Scientific/Engineering :: Artificial Intelligence'
```
later all this shall be caught by the added CI check

File: .github/workflows/ci_install-pkg.yml [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        run: |
+          # this is just a hotfix because of Win cannot install it directly
+          pip install -r requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
+
+      - name: Install | Uninstall package - archive
+        run: |
+          # install as archive
+          pip install dist/*.tar.gz
+          cd ..
+          python -c "import perceiverio ; print(perceiverio.__version__)"
```
Comment (krasserm) [inline .github/workflows/ci_install-pkg.yml:71]: Sorry if I wasn't clear enough previously. I meant import pkg `perceiverio` should be renamed to `perceiver`. OK to keep name `perceiver-io` for install pkg.

File: .github/workflows/ci_install-pkg.yml [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+          pip uninstall -y perceiver-io
+
+      - name: Install | Uninstall package - wheel
+        run: |
+          # install as wheel
+          pip install dist/*.whl
+          cd ..
+          python -c "import perceiverio ; print(perceiverio.__version__)"
```
Comment (krasserm) [inline .github/workflows/ci_install-pkg.yml:79]: Same here ...

File: .github/workflows/ci_install-pkg.yml (hunk 12 above)
Comment (Borda) [reply .github/workflows/ci_install-pkg.yml:71]:
```suggestion
          python -c "import perceiver ; print(perceiver.__version__)"
```

File: .github/workflows/ci_install-pkg.yml (hunk 13 above)
Comment (Borda) [reply .github/workflows/ci_install-pkg.yml:79]:
```suggestion
          python -c "import perceiver ; print(perceiver.__version__)"
```

**Comment by Borda** [comment]
@krasserm you can see the ci in action over here: https://github.com/Borda/perceiver-io/actions/runs/1662946314

**Review by krasserm — APPROVED** [review]
LGTM, thanks again!

**Pull request merged**